### PR TITLE
fix(ZNTA(2003): WebDriver test for alpha editor

### PR DESCRIPTION
**Pull request by djansen-redhat**
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-2003

Required setting context path to root. Adds a function to
close windows other than the first on test completion.

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/368)

### CODE REVIEW

**Comment by davidmason** [comment]
Reviewed 1 of 7 files at r1.
Review status: 1 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/functional-test/pom.xml, line 60 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqgNKviiCZNkXoFe5g:-KlqgNKviiCZNkXoFe5h:b2duyrc) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/pom.xml#L60)):*
> ```XML
>     <mysql.lower_case_table_names>0</mysql.lower_case_table_names>
> 
>     <zanata.instance.url>http://${cargo.host}:${cargo.servlet.port}/${context.path}</zanata.instance.url>
> ```

Is `context.path` used anywhere else? If not, we could just delete it. Perhaps it is around so it can be overridden or used later?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by davidmason** [comment]
Reviewed 1 of 7 files at r1.
Review status: 2 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 565 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqglbwfJyxhM6BIKuH:-KlqglbwfJyxhM6BIKuI:bqol4k) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L565)):*
> ```Java
> 
>     public String getMainWindowHandle() {
>         return Lists.newArrayList(driver.getWindowHandles()).get(0);
> ```

could make this `return getAllWindowHandles().get(0)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

File: server/functional-test/src/main/java/org/zanata/page/AbstractPage.java [hunk 1]
```diff
@@ -22,6 +22,8 @@
 
 import java.util.List;
 import java.util.Set;
+
+import org.assertj.core.util.Lists;
```
Comment (seanf) [inline server/functional-test/src/main/java/org/zanata/page/AbstractPage.java]: There's no real harm in using assertj's Lists class, but if you get in the habit of using, say, the Lists class from Guava, you can use it any code (eg non-test code, client code). Or if an immutable list is all you need, you can use Guava's ImmutableList.of(...) method.

File: server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    private By headerElement = new By.ById("editor-header");
+    private By transUnitText = new By.ByClassName("TransUnit-text");
+
+    public ReactEditorPage(WebDriver driver) {
+        super(driver);
+    }
+
+    @Override
+    protected int getExpectedBackgroundRequests() {
```
Comment (seanf) [inline server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java]: You should add a comment explaining why this is not 0, even if it's just:

    // TODO find out why we always have a background request

File: server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        log.info("Switching to new window (from {})", getDriver().getWindowHandle());
+        waitForAMoment().until((Predicate<WebDriver>) input ->
+                getAllWindowHandles().size() > 1);
+        getDriver().switchTo().window(getAllWindowHandles().get(1));
+        log.info("Window: {}", getDriver().getWindowHandle());
+        return new ReactEditorPage(getDriver());
+    }
+
+    public boolean isAlphaEditor() {
+        log.info("Query is alpha editor visible");
```
Comment (seanf) [inline server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java]: Weren't we going to call it the React editor? :-)

File: server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    // Test that the alpha button works
+    @Test
+    public void showAlphaEditor() {
+        new LoginWorkFlow()
+                .signIn("admin", "admin");
+        ReactEditorPage reactEditorPage = new BasicWorkFlow()
+                .goToEditor("about-fedora", "master", "fr", "About_Fedora")
+                .pressAlphaEditorButton()
+                .switchToEditorWindow();
```
Comment (seanf) [inline server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java]: Wait, are we really opening the React editor in a new browser tab? Ouch.

File: server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java [hunk 5]
```diff
@@ -59,6 +62,18 @@ protected void before() throws Throwable {
 
         @Override
         protected void after() {
+            // Close all windows other than the original one
+            WebDriver driverRef = WebDriverFactory.INSTANCE.getDriver();
+            List<String> windows = Lists.newArrayList(driverRef.getWindowHandles());
+            String mainWindowHandle = windows.get(0);
+            for (String windowHandle : windows) {
+                if (!windowHandle.equals(mainWindowHandle)) {
```
Comment (seanf) [inline server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java]: Couldn't you just loop i from 1 (not the usual 0) while i < windows.size()?

    for (int i = 1; i < windows.size(), i++) {

**Comment by djansen-redhat** [comment]
*[server/functional-test/pom.xml, line 60 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqgNKviiCZNkXoFe5g:-KlqrH4_bbxoDtVy2faq:buo9zvp) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/pom.xml#L60)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

Is `context.path` used anywhere else? If not, we could just delete it. Perhaps it is around so it can be overridden or used later?
</blockquote></details>

I guess it can be left there to change the path? I should put a comment on the variable to indicate this (and that a trailing / is required)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by davidmason** [comment]
Reviewed 5 of 7 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 26 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUH-r1-26:-KlqsK10QXT4kBCU7iHV:b2jmcla) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L26)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

There's no real harm in using assertj's Lists class, but if you get in the habit of using, say, the Lists class from Guava, you can use it any code (eg non-test code, client code). Or if an immutable list is all you need, you can use Guava's ImmutableList.of(...) method.
</blockquote></details>

I'm only using it to convert from a Set - is there a better way?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 565 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqglbwfJyxhM6BIKuH:-KlqsaNVIrVwM-TcaxTj:b-73m6vw) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L565)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

could make this `return getAllWindowHandles().get(0)`
</blockquote></details>

Nice catch!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
*[server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java, line 48 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUM-r1-48:-KlqtJLa_MQL3skV6B3n:bkk8hxs) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java#L48)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Wait, are we really opening the React editor in a new browser tab? Ouch.
</blockquote></details>

Yes we are. I'm on the fence as to whether that is good or not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
*[server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java, line 70 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUK-r1-70:-Klqtj4RfBs4lficjDxK:bwidv74) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java#L70)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Couldn't you just loop i from 1 (not the usual 0) while i < windows.size()?

    for (int i = 1; i < windows.size(), i++) {
</blockquote></details>

I did have that originally, but I liked the get name -> don't close if it's the wrong name check. Do you think this is too verbose?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java, line 48 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUL-r1-48:-Klqu8qWQpjTthBU2Wgu:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java#L48)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

You should add a comment explaining why this is not 0, even if it's just:

    // TODO find out why we always have a background request
</blockquote></details>

Done.

---

*[server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java, line 62 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUL-r1-62:-KlquBGL8-FT5NZKKGtk:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java#L62)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Weren't we going to call it the React editor? :-)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 26 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUH-r1-26:-KlqyPlyTybPF8E7uVtP:br91vxw) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L26)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

I'm only using it to convert from a Set - is there a better way?
</blockquote></details>

Not better per se, but you could just use `com.google.common.collect.Lists#newArrayList(java.lang.Iterable<? extends E>)`

I always get a bit suspicious when I see random Lists classes imported, instead of the usual Guava Lists. There's one in TestNG that used to creep in, and I think JCommander had one too.

Wait, it's a Set? See my comment below.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 565 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-Klr-QN1eZc5dkG8PeTh:-Klr-QN1eZc5dkG8PeTi:btpegs3) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L565)):*
> ```Java
> 
>     public String getMainWindowHandle() {
>         return Lists.newArrayList(driver.getWindowHandles()).get(0);
> ```

if `driver.getWindowHandles()` returns a Set, you may have just selected one of them at random, ie not necessarily the main window. A Java Set makes no general guarantees about the order of iteration.

It would be simpler if we just got rid of the silly pop-up when opening the React editor. It's not user-friendly anyway.

---

*[server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java, line 70 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUK-r1-70:-KlqzFG2KckNTQNufQDC:b4hv44n) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java#L70)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

I did have that originally, but I liked the get name -> don't close if it's the wrong name check. Do you think this is too verbose?
</blockquote></details>

A little. In fact, you could just use 

    List<String> windows = Lists.newArrayList(driverRef.getWindowHandles());
    windows.remove(0)
    for (String window: windows) {
        ...

or you could try a more functional version with Javaslang (the aliasing of the class `List` is a nuisance):

    javaslang.collection.List<String> windows = javaslang.collection.List.ofAll(driverRef.getWindowHandles());
    for (String window: windows.drop(1)) {
        ...

But none of that will work reliably if `driverRef.getWindowHandles()` is returning an unordered Set.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[server/functional-test/pom.xml, line 60 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqgNKviiCZNkXoFe5g:-Klr1_G8xdfEkCk8hw5S:b-20g4wl) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/pom.xml#L60)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

I guess it can be left there to change the path? I should put a comment on the variable to indicate this (and that a trailing / is required)
</blockquote></details>

:+1 to the comment.

We definitely want the option to test with alternative context paths when required, so we should keep it around. 

---

*[server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java, line 48 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUM-r1-48:-Klr1ssoQ17sXCbBGgxm:btqb4aj) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java#L48)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

Yes we are. I'm on the fence as to whether that is good or not.
</blockquote></details>

I don't think this is one of those 0.0000000001% of cases where it's justified to open a new browser tab unasked.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[server/functional-test/pom.xml, line 60 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqgNKviiCZNkXoFe5g:-KlrWiYNN3-VfsrSTqOL:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/pom.xml#L60)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

:+1 to the comment.

We definitely want the option to test with alternative context paths when required, so we should keep it around. 
</blockquote></details>

Done.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 565 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-Klr-QN1eZc5dkG8PeTh:-KlrWmoOeS09nWKyisH2:b-juxy8e) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L565)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

if `driver.getWindowHandles()` returns a Set, you may have just selected one of them at random, ie not necessarily the main window. A Java Set makes no general guarantees about the order of iteration.

It would be simpler if we just got rid of the silly pop-up when opening the React editor. It's not user-friendly anyway.
</blockquote></details>

I've made it so the test setup remembers the main window handle and closes others by name at the end.

---

*[server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java, line 48 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlqdCoCLAEDeXbR7pUM-r1-48:-KlrX2W-EekL01toqzJI:b-ybvnu8) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/test/java/org/zanata/feature/editor/EditorAlphaTest.java#L48)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

I don't think this is one of those 0.0000000001% of cases where it's justified to open a new browser tab unasked.
</blockquote></details>

Well, we can debate that outside of this task ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by seanf** [comment]
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 565 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-Klr-QN1eZc5dkG8PeTh:-Klr1QSveBLi5MjWzT1D:b-bmzln8) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L565)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

I've made it so the test setup remembers the main window handle and closes others by name at the end.
</blockquote></details>

That's much better. You should remove this method though, since it can't guarantee to do what it promises.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 569 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulS04zT244fl_NXZ1:-KlulS04zT244fl_NXZ2:bzgshi9) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L569)):*
> ```Java
> 
>     public List<String> getAllWindowHandles() {
>         return Lists.newArrayList(driver.getWindowHandles());
> ```

I don't think it's a good idea to convert a Set to a List in this case, because you're adding a false sense of order to an unordered collection.

---

*[server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java, line 57 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulfcrY6drUyJ9hwKN:-KlulfcrY6drUyJ9hwKO:by6z4zd) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java#L57)):*
> ```Java
>         waitForAMoment().until((Predicate<WebDriver>) input ->
>                 getAllWindowHandles().size() > 1);
>         getDriver().switchTo().window(getAllWindowHandles().get(1));
> ```

Unless the editor window has a consistent, known name, I think you need to switch to whichever window isn't mainWindowHandle (or the old value for `getDriver().getWindowHandle()`). `getAllWindowHandles().get(1)` isn't guaranteed to be consistent if the collection is really a Set.

---

*[server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java, line 55 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-Klul6v3Zt938UF_ZX_c:-Klul6v3Zt938UF_ZX_d:b-jyofq2) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java#L55)):*
> ```Java
>     public static final int MAX_SHORT_TEST_DURATION = 180000;
>     public static final int MAX_LONG_TEST_DURATION = 600000;
>     private static String mainWindowHandle;
> ```

You shouldn't need a static field for this one.

---

*[server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java, line 110 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulByIP7mxUB5VQUyF:-KlulByIP7mxUB5VQUyG:b-8fx2u7) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java#L110)):*
> ```Java
>         zanataRestCaller.signalBeforeTest(getClass().getName(),
>                 testName.getMethodName());
>         mainWindowHandle = WebDriverFactory.INSTANCE.getDriver().getWindowHandle();
> ```

:+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by davidmason** [comment]
Reviewed 1 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java, line 57 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulfcrY6drUyJ9hwKN:-KlunqQjImix1ewoMtnZ:blp1ei7) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java#L57)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

Unless the editor window has a consistent, known name, I think you need to switch to whichever window isn't mainWindowHandle (or the old value for `getDriver().getWindowHandle()`). `getAllWindowHandles().get(1)` isn't guaranteed to be consistent if the collection is really a Set.
</blockquote></details>

It should be `{Project Name} ({versionSlug}) to {Language} - Zanata Web Translation` but it is just `Zanata` until the javascript has loaded.

Tangent: we should be able to do better than that since we know on the server what path was requested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by djansen-redhat** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 565 at r1](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-Klr-QN1eZc5dkG8PeTh:-Klus41UkGpM-M4bIvuW:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/c32d61a4a5abb5bb96e06bb6b4fa981c4cfba743/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L565)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

That's much better. You should remove this method though, since it can't guarantee to do what it promises.
</blockquote></details>

Done.

---

*[server/functional-test/src/main/java/org/zanata/page/AbstractPage.java, line 569 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulS04zT244fl_NXZ1:-KlusVIO27tRNXhH0Y8l:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/main/java/org/zanata/page/AbstractPage.java#L569)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

I don't think it's a good idea to convert a Set to a List in this case, because you're adding a false sense of order to an unordered collection.
</blockquote></details>

Done.

---

*[server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java, line 57 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulfcrY6drUyJ9hwKN:-KlusYElq0TqIIPUr3QB:b-896fix) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java#L57)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

It should be `{Project Name} ({versionSlug}) to {Language} - Zanata Web Translation` but it is just `Zanata` until the javascript has loaded.

Tangent: we should be able to do better than that since we know on the server what path was requested.
</blockquote></details>

Done.

---

*[server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java, line 55 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-Klul6v3Zt938UF_ZX_c:-Klus_jCc0aRxMbootXf:b-asz3qb) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/test/java/org/zanata/feature/testharness/ZanataTestCase.java#L55)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

You shouldn't need a static field for this one.
</blockquote></details>

It is accessed under the static ExternalResource

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368)*

**Comment by seanf** [comment]
:lgtm:

---

Reviewed 2 of 7 files at r1, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java, line 57 at r2](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-KlulfcrY6drUyJ9hwKN:-Kluwuhno5xdLhBrBlcX:b-veruq8) ([raw file](https://github.com/zanata/zanata-platform/blob/23f94cc7a5df6c294e13239a53941a1d01b80141/server/functional-test/src/main/java/org/zanata/page/editor/ReactEditorPage.java#L57)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

Done.
</blockquote></details>



@davidmason, are you talking about the window title?

Actually, from the javadoc it appears that these window handles are opaque, so they aren't the same as `document.title` *or* `window.name`.

@djansen-redhat, this looks like a more general way of switching to a new window, although your implementation should be fine: https://stackoverflow.com/a/17275462/14379

In any case, we wouldn't need all this complexity if we got rid of the pop-up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-:-KluzZ6jJ6cj9vpBBGv4:bnfp4nl)*

**Comment by seanf** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-:-Km0RcKTLW2E1tbmTiao:bnfp4nl)*

**Comment by seanf** [comment]
@davidmason Looks like reviewable has this blocked by one of your discussions.

https://reviewable.io/reviews/zanata/zanata-platform/368#-KlulfcrY6drUyJ9hwKN

**Comment by davidmason** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/368#-:-Km0ZroY1Wgn7YGs9a4X:bnfp4nl)*

**Pull request merged**